biggest PR ever... starting to transition to scipiper/remake

wdwatkins opened the pull request:
Fortunately lots of this is model setup files though 😌 , but need to keep the canonical repo more current.

@aappling-usgs and/or @ldecicco-USGS can you take a look at the remake yml files?  `1_data_s3.yml`, `1_data_s3_cleaned.yml`, `2_setup_models` and `3_assimilate_data`.

Mainly interested in correct scipiper/remake usage and conventions.  @aappling-usgs, I made most of these remake files with whisker templates in `remake_templates`.  If my approach is completely redundant with task tables, I will make a transition to that, but this seemed more doable to get the pipeline started.

jread-usgs commented:
obligatory reminder that small(er) PRs are preferred.

wdwatkins commented:
the directories without numeric prefixes are still in my non-scipiper format.

aappling-usgs commented:
i think you mentioned at stand-up that you were trying to remove the .rds files from your repo, right? or if not, i recommend it.

wdwatkins commented:
Oh, didn't notice that there was a raw data file still committed.

aappling-usgs commented:
you might also want to not commit the .csv files? e.g. 2_setup_models/meteo/nhd_120052195_driver.csv

aappling-usgs commented on 3_assimilate_data.yml:
```diff
@@ -0,0 +1,127 @@
+      - 3_assimilate_data/out/nhd_13293262.rds.ind
+  
+  3_assimilate_data/out/nhd_167122137.rds.ind:
+    command: assimilate_data(nhd_id = I("nhd_167122137"), files = I(c('1_data_s3_cleaned/out/MN_fisheries_all_temp_data_Jan2018.rds','1_data_s3_cleaned/out/MPCA_temp_data_all.rds','1_data_s3_cleaned/out/URL_Temp_Logger_2006_to_2017.rds')), state_src = I("DOW"), state_id = I(c('04003000')))
```
you can add newlines so that this command is easier to read

ldecicco-USGS replied:
I'd been wondering about new lines since starting to use `remake`....but never had the gumption to test it out. PHEW, now I know!

wdwatkins commented:
Yeah you're probably right.  I was leaving them in that directory just not deal with caching since they weren't that big, but the driver files could eventually get bulky once we scale out to more lakes.  The nml files (`2_setup_models/nml` ) are only a few kb each though.

aappling-usgs commented on remake_templates/assimilate.template:
```diff
@@ -0,0 +1,33 @@
+packages:
```
instead of creating your own template like this, the other way to create and run (with fault tolerance) a templated remake file is to use `scipiper::create_task_plan`, `scipiper::create_task_makefile`, and `scipiper::loop_tasks`. see https://github.com/GlobalHydrologyLab/watersat/blob/master/remake.yml#L93-L106, https://github.com/GlobalHydrologyLab/watersat/blob/master/1_wqdata/src/wqp.R#L171-L241, and https://github.com/GlobalHydrologyLab/watersat/blob/master/1_wqdata/src/wqp.R#L243-L249 for an example of how to use these.

aappling-usgs commented:
Agreed, they're borderline. Do whatever works for you.

aappling-usgs commented on 3_assimilate_data.yml:
```diff
@@ -0,0 +1,127 @@
+      - lib/crosswalks/rawfile_to_id_crosswalk.csv
+      - 1_data_s3_cleaned/out/MN_fisheries_all_temp_data_Jan2018.rds
+      - 1_data_s3_cleaned/out/MPCA_temp_data_all.rds
+      - 1_data_s3_cleaned/out/URL_Temp_Logger_2006_to_2017.rds
```
again recommending that you declare dependencies in the command rather than `depends` whenever possible. In this case can be done by changing
```yml
3_assimilate_data/out/nhd_120052238.rds.ind:
  command: assimilate_data(
    nhd_id = I("nhd_120052238"),
    '1_data_s3_cleaned/out/MN_fisheries_all_temp_data_Jan2018.rds.ind'
    '1_data_s3_cleaned/out/MPCA_temp_data_all.rds.ind'
    '1_data_s3_cleaned/out/URL_Temp_Logger_2006_to_2017.rds.ind',
    state_src = I("DOW"),
    state_id = I(c('69084500')))
  depends: 
      - lib/crosswalks/rawfile_to_id_crosswalk.csv
```
(using a `...` argument in the definition of `assimilate_data` to accommodate sites with only 2 files, e.g., 3_assimilate_data/out/nhd_13293262.rds.ind)

where 
```r
assimilate_data(nhd_id, ..., state_src, state_id) {
  files <- lapply(c(...), sc_retrieve)
  # and now carry on
}
```

the above assimilate_data implementation requires that you include the new targets i've suggested for `1_data_s3_cleaned.yml` above, e.g., 
```yml
1_data_s3_cleaned/out/MPCA_temp_data_all.rds:
  command: s3_get('1_data_s3_cleaned/out/MPCA_temp_data_all.rds.ind')
```
which will allow sc_retrieve to work (`sc_retreive` actually just calls `scmake` with the data_file as a target, then returns the data_file name)

aappling-usgs commented on 3_assimilate_data.yml:
```diff
@@ -0,0 +1,127 @@
+  3_assimilate_data/out/nhd_4817675.rds.ind:
+    command: assimilate_data(nhd_id = I("nhd_4817675"), files = I(c('1_data_s3_cleaned/out/MN_fisheries_all_temp_data_Jan2018.rds','1_data_s3_cleaned/out/MPCA_temp_data_all.rds','1_data_s3_cleaned/out/URL_Temp_Logger_2006_to_2017.rds')), state_src = I("DOW"), state_id = I(c('11020300','11020301')))
+    depends: 
+      - lib/crosswalks/rawfile_to_id_crosswalk.csv
```
I don't understand how `lib/crosswalks/rawfile_to_id_crosswalk.csv` is getting used in assimilate_data or is otherwise a dependency of it. can you explain?

wdwatkins replied:
That is my solution to dealing with the fact that many of the raw files contain data for multiple lakes.  We do know what data source each file comes from though.  So for each lake the `assimilate_data` function looks at the `lakes_master` crosswalk for what state IDs are associated with an NHD lake, then goes to `rawfile_to_id_crosswalk` to see what raw files contain data for that state id's agency and checks them all for data for that site.  This does mean there is potentially some unnecessary searching, but at least right now this isn't expensive and seemed simpler than trying to generate a table of IDs in all the raw files.  

I am using "state agency" a bit loosely here, since there is data from orgs like UW, but basically they all have their own ID systems, which we need to be able to translate between, and be prepared for new ones.

wdwatkins replied:
That is how it works conceptually, but remake will actually check the hash `lib/crosswalks/rawfile_to_id_crosswalk.csv` even though it isn't created by remake, right?  This should rerun if new data files are added, since we don't know which IDs they might contain data for.

aappling-usgs replied:
OK, I think i'm starting to see. But `lakes_master` and `rawfile_to_id_crosswalk` don't actually get used in assimilate_data(), just in render_assimilate.R, right? And then render_assimilate.R uses the the contents of lib/crosswalks/rawfile_to_id_crosswalk.csv to determine the specifics of the recipes in this file (`state_src`? `files`? the `depends:` lists?)?

Are there any changes that could be made to rawfile_to_id_crosswalk.csv that would (1) cause targets in 3_assimilate_data.yml to go out of date and yet (2) if you reran render_assimilate, the text of the recipes for the out-of-date targets in 3_assimilate_data.yml wouldn't change? As I understand it so far, the answer is no. And if I have that right, then I think the ideal situation would be for this file (3_assimilate_data.yml) to itself be a target in another remake file, and to have the `depends:` of that target include `lib/crosswalks/rawfile_to_id_crosswalk.csv`. As things currently stand, changes to `lib/crosswalks/rawfile_to_id_crosswalk.csv` might help you notice the problem (because targets such as `3_assimilate_data/out/nhd_4817675.rds.ind` would be rerunning without producing anything new), but they wouldn't directly cause this .yml file to get updated and therefore wouldn't bring the out-of-date targets up to date. You'd have to recognize the issue and remember to manually rerun render_assimilate.

So: My first suggestion is to use a master remake file responsible for creating this file and the others, so that you can formalize the dependency on rawfile_to_id_crosswalk within that master file. But that'd take some work, and you might be content leaving that master-level orchestration to the data analysts (e.g. you), so my second suggestion is just to take the dependency on lib/crosswalks/rawfile_to_id_crosswalk.csv out of this file, whether or not you follow my first suggestion. And then remember to run render_assimilate each time you edit rawfile_to_id_crosswalk.csv.

wdwatkins replied:
Yeah, I think you are right.  I think this should be codified if possible.  I'm seeing better how task tables work, so I think I might as well use them here too.

aappling-usgs commented on 1_data_s3_cleaned.yml:
```diff
@@ -0,0 +1,48 @@
+  1_data_s3_cleaned/out/MPCA_temp_data_all.rds.ind:
+    command: parse_MPCA_temp_data_all()
+    depends:
+      - 1_data_s3/out/MPCA_temp_data_all.tsv.ind
```
after you've done that, i recommend creating a second target recipe for each of the files you're going to pass to assimilate_data (see suggestions below for the remake file that calls that function).
```
1_data_s3_cleaned/out/MPCA_temp_data_all.rds:
  command: s3_get('1_data_s3_cleaned/out/MPCA_temp_data_all.rds.ind')
```
where other remake targets should depend on the rds.ind file rather than depending on the .rds file directly, but they'll be able to access the .rds file within the called functions by calling
```
datafile <- sc_retrieve(indfile)
dat <- readRDS(datafile)
```
example below for assimilate_data.

wdwatkins commented on 1_data_s3.yml:
```diff
@@ -0,0 +1,47 @@
+    command: s3_get(target_name)
+
+  1_data_s3/out/MPCA_temp_data_all.tsv.ind:
+    command: s3_get(target_name)
```
I did this because of the chicken and egg problem with receiving data files from collaborators,  since a file ultimately has to be added by hand, and I was just manually creating the indicator files, but I hadn't thought of just doing a `s3_confirm_posted` to start the chain.  I don't think it has caused any remake errors though.

Although while typing this I remembered that the mendota data is publicly available, so I need to build in a step that just does a dowload.file to grab those.

aappling-usgs replied:
that explains how it was working :+1:. I'd still advocate for using `sc_confirm_posted` in the remake file to create the local .ind file, though - you just don't add that target until the corresponding data file has been put on S3.

jread-usgs commented on 1_data_s3_cleaned/src/parse_dow_files.R:
```diff
@@ -0,0 +1,47 @@
+  #keep noon measurements to downsample
+  df_clean <- df %>% mutate(DOW = "04003501", 
+                            temp = 5/9*(WaterTempF - 32),
+                            Depth = 11/3.28, 
```
why isn't this 10/3.28?

jread-usgs replied:
oh, I see`'State Waters - 11 feet'`

wdwatkins replied:
explainer file says:

wdwatkins replied:
what would 10 be from?

jread-usgs replied:
typo. Was just seeing the 3.28 and thinking that was the ft m conversion that was needed and didn't see why 11 would be in there.

wdwatkins commented:
new commits fix #8

aappling-usgs commented on 1_data_s3/out/MPCA_temp_data_all.tsv.ind:
```diff
@@ -1,2 +1,2 @@
-hash: 56a540b821c1c2966b8276036f33c457
+hash: 3bbd4c19ef247e7006ee5f3553e98cc6-43
```
interesting, i don't think i've ever seen dashes in hashes before.

wdwatkins replied:
yeah, I'm didn't thing those files has changed at all either.

wdwatkins commented:
@aappling-usgs if you want to take another look, I think I've addressed all of your comments.  I am ready to merge if this looks good, and then will get back into making reasonable size PRs...

aappling-usgs commented:
@wdwatkins thanks for the update, looks like you've done a lot :+1:. no time for another few hours but will try to take another look later today

aappling-usgs commented on 1_data_download.yml:
```diff
@@ -0,0 +1,30 @@
+    command: download_and_indicate(remote_ind = target_name, local_file = I("1_data_download/out/mendota_daily_buoy.csv"),
+              url = I("https://lter.limnology.wisc.edu/file/12861/download?token=u7og1mMGAJ9g41O1E-tiHMK8THR9gPZ6u7WpZ1HLD9U"))
+              
+  1_data_s3/out/mendota_temps_long.csv.ind:
```
You have this exact same target in 1_data_s3.yml. I assume that's by design, and I see that the two yml files don't reference one another, so it's probably OK, but I want to point out that it introduces ambiguity - if you see this .ind file in a directory, do you know what it means? If it means the exact same thing regardless of how it was built (from this yml or the other one), then was it really necessary to define its target in both places? I thiiiiink it's all fine in practice, but I also feel a vague uneasiness that I hope you share and are keeping an eye on.

wdwatkins replied:
Hmm, kind of, hadn't thought of that though.  The file that is downloaded in `1_data_download` is the same as the file that is on S3, so I was just trying to avoid an unnecessary trip to S3 and back.  Is there a better way to do that?  I suppose these files are just a few megabytes though, so it wouldnt be a big deal to send them over the wire.

aappling-usgs replied:
Good idea to try to avoid unnecessary downloads, but it's probably not working as you intend. Simply creating an .ind file isn't enough for `remake`, which also writes its own file documenting the command and inputs used to create that file. This means that two different .yml files that create the .ind file in two slightly different ways are at odds - if you built the .ind file with one .yml file the first time and then try to rebuild it with the second .yml file, you might hope that it'll just say OK and not build, but you'll be disappointed - it'll get rebuilt. And then if you build it with the first .ind file again, it'll get rebuilt again, and so on.

Fortunately, the overall system probably is working _almost_ as you intend. `scipiper` bypasses redundant data downloads within `s3_get`, here: https://github.com/USGS-R/scipiper/blob/master/R/s3.R#L154, so it's actually not very costly to call `s3_get` when you already have the data file locally. It's just a quick query to S3 to get the hash of the remotely stored file.

As I said before, I think what you already have should work OK - the only flaws are possible rebuilds of that .ind file and the ambiguity about which .yml file created the .ind file. But if you wanted to address those minor flaws, I think the way to do it would be to merge 1_data_download.yml and 1_data_s3.yml. You could do either of these:
(a) replace the recipes for these three .ind files in 1_data_s3.yml with their recipes from 1_data_download.yml, then delete 1_data_download.yml. 
(b) remove the three redundant .ind recipes from 1_data_s3.yml, and from that .yml file include 1_data_download.yml. 
With both of these options, because scipiper + git allows you to share remake build status among collaborators, these solutions should still allow just one analyst to do the original download; everybody else just uses the .ind file (plus the remake/scipiper build file) to know that the file is available to download from S3.

wdwatkins replied:
Ok, I think I will go with (b).  I'm going to go ahead and merge this and include that with my next PR, since I accidentally started it on a different branch.

aappling-usgs commented on remake.yml:
```diff
@@ -0,0 +1,73 @@
+#first render the remake file, then run it
+#need to functionalize render
```
looks like this to-do item can be crossed off?
